File: Coding_Questions/JavaScript/MinHeightBST.js
// ---Min Height BST---

class BST {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }

  insert(value) {
    let currentNode = this;

    while (true) {
      if (value < currentNode.value) {
        if (currentNode.left === null) {
          currentNode.left = new BST(value);
          break;
        } else {
          currentNode = currentNode.left;
        }
      } else {
        if (currentNode.right === null) {
          currentNode.right = new BST(value);
          break;
        } else {
          currentNode = currentNode.right;
        }
      }
    }

    return this;
  }
}

function constructMinHeightBst(array, startIndex, endIndex) {
  if (endIndex < startIndex) {
    return null;
  }

  const middleIndex = Math.floor((startIndex + endIndex) / 2);
  const bst = new BST(array[middleIndex]);

  bst.left = constructMinHeightBst(array, startIndex, middleIndex - 1);
  bst.right = constructMinHeightBst(array, middleIndex + 1, endIndex);

  return bst;
}

function minHeightBst(array) {
  return constructMinHeightBst(array, 0, array.length - 1);
}

File: Coding_Questions/C++/MinHeightBST.cpp
// ---Min Height BST---

#include <vector>
#include <cmath>


class BST {
public:
    int value;
    BST *left;
    BST *right;

    explicit BST(int value) {
        this->value = value;
        left = nullptr;
        right = nullptr;
    }

    void insert(int val) {
        BST *current_node = this;

        for (;;) {
            if (value < current_node->value) {
                if (current_node->left == nullptr) {
                    current_node->left = new BST(val);
                    break;
                } else {
                    current_node = current_node->left;
                }
            } else {
                if (current_node->right == nullptr) {
                    current_node->right = new BST(val);
                    break;
                } else {
                    current_node = current_node->right;
                }
            }
        }
    }
};


BST *construct_min_height_bst(std::vector<int> array_sub, int start_index, int end_index) {
    if (end_index < start_index) {
        return nullptr;
    }

    int middle_index = floor((end_index - start_index) / 2);

    BST *bst = new BST(array_sub[middle_index]);

    bst->left = construct_min_height_bst(array_sub, start_index, middle_index - 1);
    bst->right = construct_min_height_bst(array_sub, middle_index + 1, end_index);

    return bst;
}


BST *min_height_bst(const std::vector<int> &array) {
    return construct_min_height_bst(array, 0, array.size() - 1);
}